src/main/java/fi/aalto/cs/apluscourses/utils/CoursesClient.java

```diff
@@ -243,13 +249,36 @@ private static void consumeResponse(@NotNull HttpUriRequest request,
    * does nothing.
    */
   private static void requireSuccessStatusCode(@NotNull HttpResponse response)
-      throws UnexpectedResponseException {
+      throws IOException {
     int statusCode = response.getStatusLine().getStatusCode();
-    if (statusCode == 401 || statusCode == 403) { // TODO: should 403 be removed from this check?
-      throw new InvalidAuthenticationException(response, "Invalid authentication");
-    } else if (statusCode < 200 || statusCode >= 300) {
-      throw new UnexpectedResponseException(response, "Status code doesn't indicate success");
+    if (statusCode >= 200 && statusCode < 300) {
+      return;
+    }
+
+    String details;
+    try {
+      HttpEntity entity = response.getEntity();
+      if (entity == null) {
+        details = "" + statusCode + " " + response.getStatusLine().getReasonPhrase();
+      } else {
+        // Two possibilities: {"errors": [...]} and {"detail": "..."}
+        JSONObject json = new JSONObject(new JSONTokener(
+            new ByteArrayInputStream(EntityUtils.toByteArray(entity))
+        ));
+        details = json.optString("detail");
+        if (details == null || details.trim().isEmpty()) {
+          details = String.join(",", json
+              .getJSONArray("errors")
+              .toList()
+              .stream()
+              .map(obj -> obj.toString())
+              .collect(Collectors.toList()));
+        }
+      }
+    } catch (JSONException e) {
+      details = "" + statusCode + " " + response.getStatusLine().getReasonPhrase();
     }
```

Comment:
Nit: Here, the value of `details` is resolved with quite a complex if-else chain. It could perhaps be a bit less dangerous-looking, if it was organized as a private method where one could use return statements instead of assignments (making it clear that `details` is always assigned exactly once).
```java
private String resolveErrorMessage(HttpResponse response) {
  // ...
  if (statusCode >= 200 && statusCode < 300) {
    return null;
  }
  // ...
  if (entity == null) {
    return "" + statusCode + " " + response.getStatusLine().getReasonPhrase();
  }
  // ...
  try {
    // ...
  } catch (JSONException e) {
    // this could also be inside a separate function to avoid code duplication
    return "" + statusCode + " " + response.getStatusLine().getReasonPhrase();
  }
  // ...
  if (details != null && !details.trim().isEmpty()) {
    return details;
  }
  return String.join(...);
}

// ...
errorMsg = resolveErrorMessage(response);
```

This is a very minor issue, though, or merely an idea how to get rid of nested `if`s and `try`s.

Reply:
That's a good idea, this is quite messy ATM, so I'll implement it that way